docs/design/backup_and_restore.md

```diff
@@ -0,0 +1,202 @@
+
+Changing the following fields will require running the modify command:
+
+* Backup URL parameters
+* Snapshot time
+
+#### Toggle Backup Paused
+
+In the `Toggle Backup Paused` stage, we will pause or resume the backup, if necessary. If the desired state is `paused` and the backup is not paused, this will run the `fdbbackup pause` command. If the desired state is `running` and the backup is paused, this will run the `fdbbackup resume` command.
+
+### Reconciling Restores
+
+The restore reconciler will run the following stages:
+
+* Toggle restore running
+* Check restore status
+
+We will use the `fdbrestore` tool to modify the restore state. We will use the restore status that is found within the database status to determine the current state of the backup.
+
+#### Toggle Restore Running
+
+In the `Toggle Restore Running` stage, we will start or stop the restore, if necessary. If the desired state is `stopped`, and there is a restore running on the configured tag, this will run the `fdbrestore abort` command. If the desired state is `running`, and there is no restore running on the configured tag, this will run the `fdbrestore start` command.
+
+#### Check Restore Status
+
+In the `Check Restore Status` stage, we will compare the running state and validate the following things:
+
+* The state of the restore matches the configured state in the spec
+* If the restore is configured to be running, it has completed
+
+If any of these checks fail, the restore will be marked as unreconciled.
+
+### Expiring and Deleting Backups
+
+In order to prevent backups from growing without bound, we have configuration options to expire backup data older than a certain age. This will be done through the `fdbbackup expire` command. Because this command blocks, and requires non-trivial resources to run, this will be run in a separate pod from the main controller. The operator will maintain an in-memory queue of backup expiration jobs, with one job for each backup. On start-up, it will populate the queue, and on reconciliation it will ensure that the backup has an entry in the expiration queue.  The operator will have a limit for the concurrency of backup expiration pods, with a default of 1. The operator will have a dedicated goroutine for monitoring this queue and creating pods. If the number of active backup expiration pods is less than the concurrency limit, the queueing goroutine will pop the first job off the queue, create a pod for running that backup expiration, and put a new entry on the end of the queue. If the number of active backup expiration pods is greater than or equal to the concurrency limit, this goroutine will sleep for 10 minutes and try again. It will also check to make sure that there is only one backup expiration running for a given backup at once. Pods that have started termination will not count against this concurrency limit.
```

Comment:
Nit:  "to expire backup data older than a certain age" isn't quite right.  Should this read "to expire backup data that is no longer needed to recover to any version that is newer than the expiry date"?

It could make a difference if there are retention rules that force us to delete data that was deleted prior to a certain date, but when doing so would leave us with an incomplete backup.

Reply:
My understanding of the backup expire command is that it is somewhere between those two ideas: the semantics of the command are to delete data older than a certain age, but it has safety checks to prevent deleting data that is needed to restore to a certain point. The point in time for that safety check is provided in a different parameter.